### ISSUE TITLE
Bug: Installing transitive dependencies for pod only used in specific build configuration

### ISSUE THREAD

**Issue by rm98** [CocoaPods/CocoaPods#10474]
* [x] I've read and understood the [*CONTRIBUTING* guidelines and have done my best effort to follow](https://github.com/CocoaPods/CocoaPods/blob/master/CONTRIBUTING.md).

# Report

## What did you do?

Included a pod only for a specific configuration using `pod '<name>', '~> <version>', :configurations => ['Debug']`, then ran `pod install`.

The `Podfile` structure:
```
platform :ios, '14.4'

target 'CocoapodsTransitiveDependencyBug' do
  use_frameworks!

  pod 'TAReactorKitHelper', '~> 0.2', :configurations => ['Debug']
end
```

The `TAReactorKitHelper.podspec` contains a dependency, using `spec.dependency`.  
 [link to podspec](https://github.com/tailoredmedia/TAReactorKitHelper/blob/master/TAReactorKitHelper.podspec)

## What did you expect to happen?

The and all its transitive dependencies (that are not required by any other pod) should be only installed for the `Debug` configuration.

The `Pods-CocoapodsTransitiveDependencyBug-frameworks.sh` should contain the following section:
```sh
if [[ "$CONFIGURATION" == "Debug" ]]; then
  install_framework "${BUILT_PRODUCTS_DIR}/ReactorKit/ReactorKit.framework"
  install_framework "${BUILT_PRODUCTS_DIR}/RxSwift/RxSwift.framework"
  install_framework "${BUILT_PRODUCTS_DIR}/TAReactorKitHelper/TAReactorKitHelper.framework"
  install_framework "${BUILT_PRODUCTS_DIR}/WeakMapTable/WeakMapTable.framework"
fi
if [[ "$CONFIGURATION" == "Release" ]]; then
fi
```

## What happened instead?

In the `-frameworks.sh` file that is executed in the `[CP] Embed Pods Frameworks` build phase, the frameworks are installed.
There is an `if` check for `"$CONFIGURATION" == "Debug"`, and `"$CONFIGURATION" == "Release"`.

The pod that is named in the `Podfile` is handled correctly: it is only included in the `$CONFIGURATION" == "Debug"` if branch.  
The transitive dependencies are handled incorrectly (in my opinion): they are included in BOTH configurations, `Debug` and `Release`, although the project only uses one pod that has these as transitive dependencies, and this pod is only included for `$CONFIGURATION" == "Debug"`.

In `Pods-CocoapodsTransitiveDependencyBug-frameworks.sh`, the section that decides which frameworks to install based on `$CONFIGURATION` actually looks like this:
```sh
if [[ "$CONFIGURATION" == "Debug" ]]; then
  install_framework "${BUILT_PRODUCTS_DIR}/ReactorKit/ReactorKit.framework"
  install_framework "${BUILT_PRODUCTS_DIR}/RxSwift/RxSwift.framework"
  install_framework "${BUILT_PRODUCTS_DIR}/TAReactorKitHelper/TAReactorKitHelper.framework"
  install_framework "${BUILT_PRODUCTS_DIR}/WeakMapTable/WeakMapTable.framework"
fi
if [[ "$CONFIGURATION" == "Release" ]]; then
  install_framework "${BUILT_PRODUCTS_DIR}/ReactorKit/ReactorKit.framework"
  install_framework "${BUILT_PRODUCTS_DIR}/RxSwift/RxSwift.framework"
  install_framework "${BUILT_PRODUCTS_DIR}/WeakMapTable/WeakMapTable.framework"
fi
```

### Stack

```
   CocoaPods : 1.10.1
        Ruby : ruby 2.6.5p114 (2019-10-01 revision 67812) [x86_64-darwin19]
    RubyGems : 3.0.3
        Host : macOS 11.2 (20D64)
       Xcode : 12.4 (12D4e)
         Git : git version 2.24.3 (Apple Git-128)
Ruby lib dir : /Users/<censored>/.rbenv/versions/2.6.5/lib
Repositories : <censored>
```

### Installation Source

```
Executable Path: /Users/<censored>/.rbenv/versions/2.6.5/bin/pod
```

### Plugins

```
cocoapods-deintegrate : 1.0.4
cocoapods-plugins     : 1.0.0
cocoapods-search      : 1.0.0
cocoapods-trunk       : 1.5.0
cocoapods-try         : 1.2.0
```

### Podfile

```ruby
platform :ios, '14.4'

target 'CocoapodsTransitiveDependencyBug' do
  use_frameworks!

  pod 'TAReactorKitHelper', '~> 0.2', :configurations => ['Debug']
end
```

## Project that demonstrates the issue
[PoC Project](https://github.com/rm98/CocoapodsTransitiveDependencyBug) / [Offending File](https://github.com/rm98/CocoapodsTransitiveDependencyBug/blob/main/Pods/Target%20Support%20Files/Pods-CocoapodsTransitiveDependencyBug/Pods-CocoapodsTransitiveDependencyBug-frameworks.sh#L177)

**Comment by rm98** [comment]
When looking for the culprit, I found 
> [Note that transitive dependencies are included in all configurations and you have to manually specify build configurations for them as well in case this is not desired.](https://github.com/CocoaPods/Core/blob/6173a89b761507c76f122ec1e0da351bf1cffda3/lib/cocoapods-core/podfile/dsl.rb#L172)

Is this really the way this should work?
Or is it just complex to integrate in CocoaPods, and therefore, it is implemented this way currently?

**Comment by rm98** [comment]
Also, I [found a PR](https://github.com/CocoaPods/CocoaPods/pull/9066) that aimed to tackle it but failed to get merged because handling all edge-cases was too complex.

Therefore, I guess a PR that behavior is just in place because it is complex to integrate it in CocoaPods, and it may be changed by contributers.

**Comment by dnkoutso** [comment]
That is correct. There was an attempt that was made. For now the work around is to specify the transitive dependencies with the same configuration.
